src/main/java/software/amazon/awssdk/crt/eventstream/ClientConnectionContinuationHandler.java

```diff
@@ -0,0 +1,67 @@
+package software.amazon.awssdk.crt.eventstream;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Handler interface for responding to continuation events. It's auto closable.
+ * By default, onContinuationClosed() releases the underlying resource.
+ */
+public abstract class ClientConnectionContinuationHandler implements AutoCloseable {
+    // this gets set upon creation of the using ClientConnectionContinuation
+    protected ClientConnectionContinuation continuation;
+    private CompletableFuture<Void> closedFuture = new CompletableFuture<>();
+
+    /**
+     * Invoked when a message is received on a continuation.
+     * @param headers List of EventStream headers for the message received.
+     * @param payload Payload for the message received
+     * @param messageType message type for the message
+     * @param messageFlags message flags for the message
+     */
+    protected abstract void onContinuationMessage(final List<Header> headers,
+                          final byte[] payload, final MessageType messageType, int messageFlags);
+
+    private void onContinuationMessageShim(final byte[] headersPayload, final byte[] payload,
+                                   int messageType, int messageFlags) {
+        List<Header> headers = new ArrayList<>();
+
+        ByteBuffer headersBuffer = ByteBuffer.wrap(headersPayload);
+        while (headersBuffer.hasRemaining()) {
+            Header header = Header.fromByteBuffer(headersBuffer);
+            headers.add(header);
+        }
+
+        onContinuationMessage(headers, payload, MessageType.fromEnumValue(messageType), messageFlags);
+    }
+
+    /**
+     * By default closes the underlying resource. If you override this function, be sure to
+     * either call close() manually or invoke super.onContinuationClosed() before returning.
+     */
+    protected void onContinuationClosed() {
+        this.close();
+    }
+
+    void onContinuationClosedShim() {
+        onContinuationClosed();
+        closedFuture.complete(null);
+    }
```

Comment:
why not call `this.close()` from the `onContinuationClosedShim()`? Then you don't need to worry about the user forgetting to call it from their override.

Or are you trying to give the user control over calling close() before or after they take their custom actions?

Reply:
I'd rather the user have the ability to change the behavior if they need to (like defer close() for whatever reason). Worst case we can come back and change it if I'm wrong. close() is idempotent.